sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala

```diff
@@ -64,49 +64,57 @@ case class SortOrder(child: Expression, direction: SortDirection)
 }
 
 /**
- * An expression to generate a 64-bit long prefix used in sorting.
+ * An expression to generate a 64-bit long prefix used in sorting. If the sort must operate over
+ * null keys as well, this.nullValue can be used in place of emitted null prefixes in the sort.
  */
 case class SortPrefix(child: SortOrder) extends UnaryExpression {
 
+  val nullValue = child.child.dataType match {
+    case BooleanType | DateType | TimestampType | _: IntegralType =>
+      Long.MinValue
+    case dt: DecimalType if dt.precision - dt.scale <= Decimal.MAX_LONG_DIGITS =>
+      Long.MinValue
+    case dt: DecimalType =>
+      DoublePrefixComparator.computePrefix(Double.NegativeInfinity)
```

Comment:
It seems `dt` is not used here. Maybe this could be as below:

```scala
case _: DecimalType =>
  DoublePrefixComparator.computePrefix(Double.NegativeInfinity)
```